sourcecode/apis/contentauthor/app/Libraries/H5P/H5PLibraryAdmin.php

```diff
@@ -44,11 +44,18 @@ public function handleUpload(string $path, bool $upgradeOnly, bool $disableFileC
         // Move so core can validate the file extension.
         rename($path, $newPath);
 
-        if (!$this->validator->isValidPackage(true, $upgradeOnly)) {
-            @unlink($this->framework->getUploadedH5pPath());
+        try {
+            if (!$this->validator->isValidPackage(true, $upgradeOnly)) {
+                @unlink($this->framework->getUploadedH5pPath());
 
+                throw new InvalidH5pPackageException(
```

Comment:
This is going to be immediately caught, so `unlink()` will be called twice, and the errors passed to the constructor will be lost.